From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mikhail Aizatsky <maizatskyi@cloudflare.com>
Date: Tue, 16 Jun 2026 22:00:00 +0000
Subject: Fix ExtendedMap layout on Windows

MSVC's C++ object layout can place derived-class members into base-class
tail padding. ExtendedMap intentionally leaves kTaggedSize - 1 bytes after
bit_field_ex_ for subclasses, but Torque expects ExtendedMap itself to be
rounded up before JSInterceptorMap fields start.

Make the padding explicit in ExtendedMap's Torque and C++ layouts, and give
JSInterceptorMap's flags padding a distinct name. This keeps the C++ layout
and generated Torque offsets consistent on Windows.

There are now two padding arrays and both must be zeroed, otherwise
nondeterministic bytes end up in the heap and in snapshots.

diff --git a/src/objects/js-interceptor-map-inl.h b/src/objects/js-interceptor-map-inl.h
index e10726eb855b2573d8750573e6decaff63ad1ef4..061c6e7c53ec077f9cd3f13e50892d84881c2687 100644
--- a/src/objects/js-interceptor-map-inl.h
+++ b/src/objects/js-interceptor-map-inl.h
@@ -53,6 +53,7 @@ void JSInterceptorMap::set_supports_fast_iterable_to_list(bool value) {
 void JSInterceptorMap::init_flags_and_clear_extended_padding() {
   flags_ = 0;
   memset(extended_padding_, 0, sizeof(extended_padding_));
+  memset(flags_padding_, 0, sizeof(flags_padding_));
 }
 
 Tagged<InterceptorInfo> Map::GetNamedInterceptor() const {
diff --git a/src/objects/js-interceptor-map.h b/src/objects/js-interceptor-map.h
index 1bf77351fa028f6c7b2147827f38590aedfa0e1d..4b816042831c0c50d8171ce9de7509036e4f2272 100644
--- a/src/objects/js-interceptor-map.h
+++ b/src/objects/js-interceptor-map.h
@@ -41,7 +41,7 @@ V8_OBJECT class JSInterceptorMap : public ExtendedMap {
 
  public:
   uint8_t flags_;
-  uint8_t extended_padding_[kTaggedSize - 2];
+  uint8_t flags_padding_[kTaggedSize - 1];
   TaggedMember<InterceptorInfo> named_interceptor_;
   TaggedMember<InterceptorInfo> indexed_interceptor_;
   TaggedMember<Cell> fast_case_validity_cell_;
diff --git a/src/objects/js-interceptor-map.tq b/src/objects/js-interceptor-map.tq
index f351cb0fb0a75ab7639b56d0e4f8c0dc10e935e8..3cbe7c66e301af9dbd017598bb4985d0cbd9b051 100644
--- a/src/objects/js-interceptor-map.tq
+++ b/src/objects/js-interceptor-map.tq
@@ -9,8 +9,8 @@
 @doNotGenerateCast
 extern class JSInterceptorMap extends ExtendedMap {
   flags: uint8;
-  @ifnot(TAGGED_SIZE_8_BYTES) extended_padding[2]: uint8;
-  @if(TAGGED_SIZE_8_BYTES) extended_padding[6]: uint8;
+  @ifnot(TAGGED_SIZE_8_BYTES) flags_padding[3]: uint8;
+  @if(TAGGED_SIZE_8_BYTES) flags_padding[7]: uint8;
   named_interceptor: InterceptorInfo;
   indexed_interceptor: InterceptorInfo;
   // This prototype validity cell guards modifications of the interceptor's
diff --git a/src/objects/map.h b/src/objects/map.h
index 594d4c7c79ae2ad007163ca9881b0a4b949ca17e..f31c99af4e27c11668c37a140b167c5a9c5d9264 100644
--- a/src/objects/map.h
+++ b/src/objects/map.h
@@ -1288,7 +1288,8 @@ V8_ABSTRACT_OBJECT class ExtendedMap : public Map {
   static const int kStartOfStrongExtendedFieldsOffset;
 
   std::atomic<uint8_t> bit_field_ex_;
-  // Leaves kTaggedSize-1 unused bytes, they will be used by subclasses.
+  // Pads ExtendedMap to kTaggedSize so subclasses start at tagged alignment.
+  uint8_t extended_padding_[kTaggedSize - 1];
 } V8_OBJECT_END;
 
 constexpr int ExtendedMapSizeForKind(ExtendedMapKind kind);
diff --git a/src/objects/map.tq b/src/objects/map.tq
index cb85e7788e6a38116fc690341ca51f6288869247..687c95966d88c31b48ff3cb045c452762b2af237 100644
--- a/src/objects/map.tq
+++ b/src/objects/map.tq
@@ -116,7 +116,9 @@ bitfield struct ExtendedMapBitFields extends uint8 {
 @doNotGenerateCast
 extern class ExtendedMap extends Map {
   bit_field_ex: ExtendedMapBitFields;
-  // Leaves kTaggedSize-1 unused bytes, they will be used by subclasses.
+  // Pads ExtendedMap to kTaggedSize so subclasses start at tagged alignment.
+  @ifnot(TAGGED_SIZE_8_BYTES) extended_padding[3]: uint8;
+  @if(TAGGED_SIZE_8_BYTES) extended_padding[7]: uint8;
 }
 
 @export
